File: javascript/combograms.js
/* 
  Combograms in JavaScript.

  List all the possible words that can created given a source word.
  
  This JavaScript program was created by Hakan Kjellerstrand, hakank@gmail.com
  See also my JavaScript page: http://www.hakank.org/javascript_progs/

*/
'use strict';
// const utils = require('./js_utils.js');
const fs = require("fs");

// Allowed types
const TYPES = {
    as_many_as_wanted: 1,
    as_many_as_in_source: 2,
    plain_subwords: 3
}

// Allowed languages
const LANGUAGES = {
    "eng": 1,
    "swe": 2
}

/**
 * 
 * Create a hash from a word with keys are the characters and the
 * values are the count of each character.
 * Here is the word map of "programming":
 *   { p: 1, r: 2, o: 1, g: 2, a: 1, m: 2, i: 1, n: 1 }
 *
 * @param{string} word 
 * 
 * @returns a word hash
 * 
 */
function word_map(word) {
    let h = {};
    for(let c of word) {
        h[c] = 0;
    }
    for(let c of word) {
        h[c]++;
    }
    return h;    
}



/**
 * 
 * @param {string} word 
 * @param {string} type (as_many_as_wanted,as_many_as_in_source, plain_subwords)
 * @param {string} language ("eng", "swe")
 * 
 * @returns an array of the matched words
 * 
 */
function combogram(word,
                   lang="eng",
                   type="as_many_as_in_source",
                   must_contain="",
                   min_length=0) {

    // All wordlists are in lower case
    word = word.toLowerCase();

    console.log("word:",word, "lang:",lang, "type:", type);

    if (TYPES[type] === undefined) {
        console.log("Invalid type: ", type);
        return [];
    } 
    if (LANGUAGES[lang] === undefined ) {
        console.log("Invalid language:", lang);
        return [];
    } 

    // let wordlist = "words_lower.txt";
    let wordlist = "eng_dict.txt";
    // let wordlist = "unixdict.txt"; // English (much smaller)

    if (lang === "swe") {
        wordlist = "/home/hakank/public_html/combograms/sv_spelling_org_utf8.txt";
    }

    const words = fs.readFileSync(wordlist,"utf8").split("\n");
    console.log("number of words: " + words.length);
    // words.forEach(w=>console.log(`"${w}",`));

    
    // Word map of source word
    const h = word_map(word);

    let res = [];
    LOOP: 
    for (let w of words) {

        if (w.length === 0) {
            continue;
        }

        if (min_length > 0 && w.length < min_length) {
            continue;
        }
  
        // Check this word
        let wh = word_map(w);
        for (let c of Object.keys(wh)) {
            // Character not in source word
            if (h[c] === undefined) {
                continue LOOP;
            }

            if (must_contain.length > 0) {
                for(let mc of must_contain) {
                    if (wh[mc] === undefined) {
                        continue LOOP;
                    }
                }
            }

            if (type === "plain_subwords") {
                // Not a subword
                if (word.indexOf(w) === -1) {
                    continue LOOP;
                }

            } else if (type === "as_many_as_in_source") {
                if (wh[c] > h[c]) {
                    // Too many occurrences of this character
                    continue LOOP;
                }
            }
        }
        console.log(w);
        res.push(w);
    }
    return res;
}

// Test it
function test() {

    const t1_start = +new Date();

    let word = "programming";
    // let word = "javascript";
    // let word = "kjellerstrand";

    // Language
    // let lang = "eng";
    let lang = "swe";

    // Type
    // let type ="plain_subwords";
    // let type ="as_many_as_in_source";
    let type = "as_many_as_wanted";

    let must_contain = "";
    let min_length = 0;

    const cs = combogram(word, lang, type, must_contain, min_length);
    
    console.log("count", cs.length);

    const t1_end = +new Date();
    console.log("time (ms): ", t1_end - t1_start);

}

test();

File: julia/combograms.jl
#=
  Combograms in Julia.

  List all the possible words that can created given a source word,
  with some options:
  - as many characters as in the word, or as many as wanted 
  - plain subwords 
  - anagrams 
  - minimum length 
  - must contain certain characters 

  This program was created by Hakan Kjellerstrand, hakank@gmail.com
  See also my Julia page: http://www.hakank.org/julia/

=#


#
# Return a Dict of the elements and their occurences
# in the collection a.
#
function make_hash(a)
    d = Dict{Char,Int}()
    for e in collect(a)
        if !haskey(d,e)
            d[e] = 0
        end
        d[e] += 1
    end
    return d
end

#
# combogram(word,
#           lang,         # :eng, :swe
#           type,         # :as_many_as_in_source,:as_many_as_wanted, :plain_subwords, :anagram
#           must_contain, # character that must be in the word 
#           min_length,   # minimum length of the word 
#           )
#
function combogram(word,
                   lang=:eng,
                   type=:as_many_as_in_source,
                   must_contain="",
                   min_length=0)

    word = lowercase(word)
    println("word:$word lang:$lang type:$type must_contain:$must_contain min_length:$min_length")

    wordlist = "eng_dict.txt" # :eng
    if lang == :swe 
        # Swedish words
        wordlist = "sv_spelling_org_utf8.txt"
    end

    if must_contain != "" 
        must_contain = collect(must_contain)
    end

    words = readlines(wordlist)
    println("Number of words: $(length(words))")

    d = make_hash(word) # Characters in source word

    res = []
    cc = 0
    for w in words

        w_len = length(w)
        if w_len == 0
            continue
        end

        if min_length > 0 && w_len < min_length
            continue
        end 

        # Hash for this word
        wd = make_hash(w)

        # Anagrams
        if type == :anagram 
            if wd != d 
                continue
            end
        end

        # Plain subword
        if type == :plain_subwords
            if !occursin(w,word)
                continue
            end
        end

        good = true
        # Check all characters
        for c in keys(wd)

            # Char is not in source word
            if !haskey(d,c)
                good = false
                break
            end
            if !good continue end 

            # Has not a mandatory char
            if must_contain != "" 
                for mc in must_contain 
                    if !haskey(wd,mc)
                        good = false
                        break
                    end
                end
            end 
            if !good continue end 

            
            if type == :as_many_as_in_source 
                # Too many occurrences of this character
                if wd[c] > d[c] 
                    good = false
                    break 
                end
            end 
            if !good continue end 
            
        end

        if good 
            push!(res,w)
        end

    end
    return res
end


# 747 Swedish words
# res = combogram("programming",:swe,:as_many_as_wanted,"",0) 

# 615 English words
# res = combogram("programming",:eng,:as_many_as_wanted,"",0) 

# Plain subwords
# res = combogram("kjellerstrand",:swe,:plain_subwords,"",0) 

# Anagrams
# res = combogram("glare",:eng,:anagram,"",0) 

# min length
# Any["darkeners", "slanderer", "terrellas", "treadlers"]
# res = combogram("kjellerstrand",:eng,:as_many_as_in_source,"",9)


len = length(res)
println(res)
println("found $len words")

